In [ ]:
# This R environment comes with many helpful analytics packages installed
# It is defined by the kaggle/rstats Docker image: https://github.com/kaggle/docker-rstats
# For example, here's a helpful package to load

library(tidyverse) # metapackage of all tidyverse packages

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

list.files(path = "../input")



[1] "sales-data"

In [ ]:
# Load required libraries
library(dplyr)      # Data manipulation
library(ggplot2)    # Data visualization
library(lubridate)  # Date and time manipulation
library(readxl)     # Import spreadsheets
#Import the dataset
sales_data <- read_excel("/kaggle/input/sales-data/Sales(1).xlsx")
head(sales_data, 3)

manufact  model  sales  resale type price  engine_s horsepow wheelbas width
1 Chevrolet Metro  21.855 5.16   0     9.235 1.0       55       93.1    62.6 
2 Hyundai   Accent 41.184 5.86   0     9.699 1.5       92         NA    65.7 
3 Saturn    SL     80.620 9.20   0    10.685 1.9      100      102.4    66.4 
  length curb_wgt fuel_cap mpg Grp_Horse
1 149.4  1.895    10.3     45   1       
2 166.7  2.240    11.9     31  NA       
3 176.9     NA    12.1     33   1

**1(a) Update the missing observations using any appropriate technique. Describe the
and illustrate the chosen technique**

In [ ]:
#Check how many missing values are there
missing_values=is.na(sales_data)
sum(missing_values)
summary(sales_data)

[1] 115

   manufact            model               sales            resale     
 Length:157         Length:157         Min.   :  0.11   Min.   : 5.16  
 Class :character   Class :character   1st Qu.: 13.96   1st Qu.:11.26  
 Mode  :character   Mode  :character   Median : 29.45   Median :14.18  
                                       Mean   : 53.91   Mean   :18.07  
                                       3rd Qu.: 69.32   3rd Qu.:19.88  
                                       Max.   :540.56   Max.   :67.55  
                                       NA's   :6        NA's   :36     
      type          price           engine_s        horsepow    
 Min.   :0.00   Min.   : 9.235   Min.   :1.000   Min.   : 55.0  
 1st Qu.:0.00   1st Qu.:18.176   1st Qu.:2.300   1st Qu.:149.0  
 Median :0.00   Median :23.099   Median :3.000   Median :180.0  
 Mean   :0.26   Mean   :27.646   Mean   :3.054   Mean   :186.6  
 3rd Qu.:1.00   3rd Qu.:32.831   3rd Qu.:3.500   3rd Qu.:215.0  
 Max.   :1.00   Max.   :85.500   M

Looking at the summary statistics, we can see variables *sales*, *resale*, *type*, *price, engine_s, horsepow, wheelbas, width, length, curb_wgt, fuel_cap, mpg, Grp_Horse* all have NA values of varying amounts.

In [ ]:
#Instead of pasting 13 lines of code to carry out imputing,
   #we shall make a loop in order to make our work cleaner

# List of variables to impute
variables_to_impute <- c("type", "sales", "resale", "price",
                         "engine_s", "horsepow", "wheelbas",
                         "width", "length", "curb_wgt", "fuel_cap",
                         "mpg", "Grp_Horse")

# Loop through each variable and impute missing values with mean
for (variable in variables_to_impute) {

  sales_data[[variable]][is.na(sales_data[[variable]])] <- mean(sales_data[[variable]], na.rm = TRUE)
}
summary(sales_data)
head(sales_data, 7)

   manufact            model               sales            resale     
 Length:157         Length:157         Min.   :  0.11   Min.   : 5.16  
 Class :character   Class :character   1st Qu.: 14.35   1st Qu.:12.54  
 Mode  :character   Mode  :character   Median : 31.15   Median :17.71  
                                       Mean   : 53.91   Mean   :18.07  
                                       3rd Qu.: 67.96   3rd Qu.:18.14  
                                       Max.   :540.56   Max.   :67.55  
      type          price           engine_s        horsepow    
 Min.   :0.00   Min.   : 9.235   Min.   :1.000   Min.   : 55.0  
 1st Qu.:0.00   1st Qu.:18.575   1st Qu.:2.300   1st Qu.:150.0  
 Median :0.00   Median :24.340   Median :3.000   Median :185.0  
 Mean   :0.26   Mean   :27.646   Mean   :3.054   Mean   :186.6  
 3rd Qu.:0.26   3rd Qu.:31.930   3rd Qu.:3.500   3rd Qu.:215.0  
 Max.   :1.00   Max.   :85.500   Max.   :8.000   Max.   :450.0  
    wheelbas         width           leng

manufact  model    sales   resale type price  engine_s horsepow wheelbas
1 Chevrolet Metro     21.855  5.160 0     9.235 1.0       55       93.1000
2 Hyundai   Accent    41.184  5.860 0     9.699 1.5       92      107.4573
3 Saturn    SL        80.620  9.200 0    10.685 1.9      100      102.4000
4 Saturn    SC        24.546 10.590 0    12.535 1.9      100      102.4000
5 Acura     Integra   16.919 16.360 0    21.500 1.8      140      101.2000
6 Audi      A4        20.397 22.255 0    23.990 1.8      150      102.6000
7 Chevrolet Cavalier 145.519  9.250 0    13.260 2.2      115      104.1000
  width length curb_wgt fuel_cap mpg      Grp_Horse
1 62.6  149.4  1.895000 10.30000 45.00000 1.000000 
2 65.7  166.7  2.240000 11.90000 31.00000 3.310811 
3 66.4  176.9  3.392631 12.10000 33.00000 1.000000 
4 66.4  180.0  2.367000 12.10000 23.76093 1.000000 
5 67.3  172.4  2.639000 18.04533 28.00000 2.000000 
6 68.2  178.0  2.998000 16.40000 27.00000 2.000000 
7 67.9  180.9  2.676000 14.30000 27.00000 2.000000

Missing Values are now dealt with. These missing values were replaced by the means of corresponding columns/variables, since I thought deleting a whole 115 NAs would make further analysis innacurate

**1(b) Add two appropriate “time and repeated” variables in the data. Describe the
technique used and the variables added to the data**

In [ ]:
# Generate different years between 2013 and 2022
set.seed(123)
date = sort(as.integer (sample (2013:2022, 11, replace = TRUE)))
a = data.frame(sales_data$manufact, sales_data$model)
colnames(a) = c("manufact", "model")

a = a %>% group_by (manufact) %>% mutate (model_id = row_number()) %>% ungroup()
highest_model_no = sort (unique (a $model_id))
highest_model_no

date

year = c()
for (i in 1:nrow(a)){
  year[i] = date [a$model_id[i]]
}
year = data.frame(year)
sales_data$time = year

mnf = sort (unique (sales_data$manufact))
mnf

repeated_var = 1:30
repeated_var

sales_data = sales_data %>% mutate(mnf_ID =match(manufact, sort(unique(manufact))))
colnames(sales_data)

[1]  1  2  3  4  5  6  7  8  9 10 11

[1] 2014 2015 2015 2016 2017 2017 2018 2018 2021 2022 2022

[1] "Acura"         "Audi"          "BMW"           "Buick"        
 [5] "Cadillac"      "Chevrolet"     "Chrysler"      "Dodge"        
 [9] "Ford"          "Honda"         "Hyundai"       "Infiniti"     
[13] "Jaguar"        "Jeep"          "Lexus"         "Lincoln"      
[17] "Mercedes-Benz" "Mercury"       "Mitsubishi"    "Nissan"       
[21] "Oldsmobile"    "Plymouth"      "Pontiac"       "Porsche"      
[25] "Saab"          "Saturn"        "Subaru"        "Toyota"       
[29] "Volkswagen"    "Volvo"

[1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
[26] 26 27 28 29 30

[1] "manufact"  "model"     "sales"     "resale"    "type"      "price"    
 [7] "engine_s"  "horsepow"  "wheelbas"  "width"     "length"    "curb_wgt" 
[13] "fuel_cap"  "mpg"       "Grp_Horse" "time"      "mnf_ID"

**2. Consider the newly created data in 1(b) above**
* **(a) Make the data to be in wide format. Describe the technique used**

In [ ]:
# Convert data to wide format using pivot_wider function from tidyr
library(tidyr)
wide_data = pivot_wider(sales_data, names_from = mnf_ID, values_from = price)
print(wide_data)

# A tibble: 157 × 45
   manufact  model    sales resale  type engine_s horsepow wheelbas width length
   <chr>     <chr>    <dbl>  <dbl> <dbl>    <dbl>    <dbl>    <dbl> <dbl>  <dbl>
 1 Chevrolet Metro     21.9   5.16     0      1         55     93.1  62.6   149.
 2 Hyundai   Accent    41.2   5.86     0      1.5       92    107.   65.7   167.
 3 Saturn    SL        80.6   9.2      0      1.9      100    102.   66.4   177.
 4 Saturn    SC        24.5  10.6      0      1.9      100    102.   66.4   180 
 5 Acura     Integra   16.9  16.4      0      1.8      140    101.   67.3   172.
 6 Audi      A4        20.4  22.3      0      1.8      150    103.   68.2   178 
 7 Chevrolet Cavalier 146.    9.25     0      2.2      115    104.   67.9   181.
 8 Chevrolet Prizm     32.3   9.12     0      1.8      120     97.1  66.7   174.
 9 Chrysler  Cirrus    53.9  12.6      0      2        132    108    71     186 
10 Dodge     Neon      76.0   7.75     0      2        132    105    74.4   174.
# ℹ 147

**2(b) Make the data in 2 (a) to be in long format. Describe the technique used**

In [ ]:
# Convert data to long format using 'pivot_longer' function
long_data <- pivot_longer(wide_data, cols = starts_with("manufact"),
                          names_to = "mnf_ID", values_to = "Price")
print(long_data)


# A tibble: 157 × 46
   model    sales resale  type engine_s horsepow wheelbas width length curb_wgt
   <chr>    <dbl>  <dbl> <dbl>    <dbl>    <dbl>    <dbl> <dbl>  <dbl>    <dbl>
 1 Metro     21.9   5.16     0      1         55     93.1  62.6   149.     1.90
 2 Accent    41.2   5.86     0      1.5       92    107.   65.7   167.     2.24
 3 SL        80.6   9.2      0      1.9      100    102.   66.4   177.     3.39
 4 SC        24.5  10.6      0      1.9      100    102.   66.4   180      2.37
 5 Integra   16.9  16.4      0      1.8      140    101.   67.3   172.     2.64
 6 A4        20.4  22.3      0      1.8      150    103.   68.2   178      3.00
 7 Cavalier 146.    9.25     0      2.2      115    104.   67.9   181.     2.68
 8 Prizm     32.3   9.12     0      1.8      120     97.1  66.7   174.     2.40
 9 Cirrus    53.9  12.6      0      2        132    108    71     186      2.91
10 Neon      76.0   7.75     0      2        132    105    74.4   174.     2.57
# ℹ 147 more rows
#

**2(C) Generate 2 variables/observations of your choice and add them to 1b**

**2(d) Provide appropriate descriptive summaries of any two variables using data in 2(a).
Also provide interpretation**

In [ ]:
summary(sales_data)
summary_sales <- summary(sales_data$sales)
summary_sales


   manufact            model               sales            resale     
 Length:157         Length:157         Min.   :  0.11   Min.   : 5.16  
 Class :character   Class :character   1st Qu.: 14.35   1st Qu.:12.54  
 Mode  :character   Mode  :character   Median : 31.15   Median :17.71  
                                       Mean   : 53.91   Mean   :18.07  
                                       3rd Qu.: 67.96   3rd Qu.:18.14  
                                       Max.   :540.56   Max.   :67.55  
      type          price           engine_s        horsepow    
 Min.   :0.00   Min.   : 9.235   Min.   :1.000   Min.   : 55.0  
 1st Qu.:0.00   1st Qu.:18.575   1st Qu.:2.300   1st Qu.:150.0  
 Median :0.00   Median :24.340   Median :3.000   Median :185.0  
 Mean   :0.26   Mean   :27.646   Mean   :3.054   Mean   :186.6  
 3rd Qu.:0.26   3rd Qu.:31.930   3rd Qu.:3.500   3rd Qu.:215.0  
 Max.   :1.00   Max.   :85.500   Max.   :8.000   Max.   :450.0  
    wheelbas         width           leng

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.11   14.35   31.15   53.91   67.96  540.56 

**Sales:**

The minimum price value is 0.11.
The first quartile (25th percentile) is 14.35.
The median (50th percentile) is 31.15.
The mean (average) price value is 53.91.
The third quartile (75th percentile) is 67.96.
The maximum price value is 540.56.
Interpretation:

Minimum: The lowest price value in the dataset is 0.11, indicating that there is at least one entry with a very low price.
Quartiles: These quartiles divide the data into four equal parts. The first quartile (14.35) represents the 25th percentile, the median (31.15) is the 50th percentile, and the third quartile (67.96) is the 75th percentile.
Mean: The average price value across all observations is 53.91. This value is influenced by the distribution of all price figures.
Maximum: The highest price value in the dataset is 540.56, indicating that there is at least one entry with a very high price.

**Price:**

Mean: The average price of cars in this column is 0.26.
Min and Max: The prices range from 0.00 to 1.00.
Interpretation: The mean value of 0.26 indicates that, on average, cars have a price around this value. The range from 0.00 to 1.00 suggests that the dataset may use a normalized or scaled representation of prices, where 1.00 could represent a maximum value or a specific unit.